File: parameters/src/node_with_params.cpp
#include "rclcpp/rclcpp.hpp"

class NodeWithParams : public rclcpp::Node
{
public:
    NodeWithParams() : Node("node_with_params")
    {
        
        // DÉCLARER LES PARAMÈTRES
        // declare_parameter(nom, valeur_par_defaut)
        // Si le paramètre n'est pas fourni au démarrage,
        // la valeur par défaut est utilisée
        this->declare_parameter("timer_period", 1.0);
        //                       |               |
        //                      nom          valeur par défaut (double)

        this->declare_parameter("message", std::string("Hello ROS2 !"));
        //                                  |
        //                     std::string obligatoire pour les strings en C++

        this->declare_parameter("max_count", 10);
        //                                    |
        //                               valeur par défaut (int)


        // LIRE LES PARAMÈTRES
        double timer_period = this->get_parameter("timer_period").as_double();
        //                                                          |
        //                                         .as_double() : convertir en double
        //                                         .as_string() : convertir en string
        //                                         .as_int()    : convertir en int
        //                                         .as_bool()   : convertir en bool

        message_  = this->get_parameter("message").as_string();
        max_count_ = this->get_parameter("max_count").as_int();
        count_ = 0;

        RCLCPP_INFO(this->get_logger(), "Parameters loaded :");
        RCLCPP_INFO(this->get_logger(), "  timer_period : %.2f", timer_period);
        RCLCPP_INFO(this->get_logger(), "  message      : %s",   message_.c_str());
        RCLCPP_INFO(this->get_logger(), "  max_count    : %d",   max_count_);


        // CRÉER LE TIMER avec la période du paramètre
        timer_ = this->create_wall_timer(
            std::chrono::duration<double>(timer_period),
            //  | utiliser la valeur du paramètre
            std::bind(&NodeWithParams::timerCallback, this)
        );
    }

private:
    void timerCallback()
    {
        if (count_ >= max_count_)
        {
            RCLCPP_WARN(this->get_logger(), "Max count reached ! Stopping...");
            timer_->cancel();   // arrêter le timer
            return;
        }

        count_++;
        RCLCPP_INFO(this->get_logger(), "[%d/%d] %s",
            count_, max_count_, message_.c_str());
    }

    rclcpp::TimerBase::SharedPtr timer_;
    std::string message_;
    int max_count_;
    int count_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<NodeWithParams>());
    rclcpp::shutdown();
    return 0;
}

File: parameters_py/parameters_py/node_with_params.py
import rclpy
from rclpy.node import Node

class NodeWithParams(Node):
    def __init__(self):
        super().__init__("node_with_params")


        # DÉCLARER LES PARAMÈTRES
        self.declare_parameter("timer_period", 1.0)
        #                       |               |
        #                      nom         valeur par défaut

        self.declare_parameter("message", "Hello ROS2 !")
        self.declare_parameter("max_count", 10)


        # LIRE LES PARAMÈTRES
        timer_period = self.get_parameter("timer_period").get_parameter_value().double_value
        #                                                  |
        #                                  get_parameter_value() : récupérer la valeur
        #                                  .double_value  : en double
        #                                  .string_value  : en string
        #                                  .integer_value : en int
        #                                  .bool_value    : en bool

        self.message_   = self.get_parameter("message").get_parameter_value().string_value
        self.max_count_ = self.get_parameter("max_count").get_parameter_value().integer_value
        self.count_ = 0

        self.get_logger().info("Parameters loaded :")
        self.get_logger().info(f"  timer_period : {timer_period}")
        self.get_logger().info(f"  message      : {self.message_}")
        self.get_logger().info(f"  max_count    : {self.max_count_}")


        # CRÉER LE TIMER avec la période du paramètre
        self.timer_ = self.create_timer(timer_period, self.timer_callback)


    # CALLBACK DU TIMER
    def timer_callback(self):
        if self.count_ >= self.max_count_:
            self.get_logger().warn("Max count reached ! Stopping...")
            self.timer_.cancel()   # arrêter le timer
            return

        self.count_ += 1
        self.get_logger().info(f"[{self.count_}/{self.max_count_}] {self.message_}")

def main(args=None):
    rclpy.init(args=args)
    node = NodeWithParams()
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == "__main__":
    main()